Hop nodes for k8s

fosterseth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Allow users to setup hop nodes for k8s deployments

- [ ] revert using devel receptor-collection before merging

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 22.0.1
```

fosterseth committed:
Add support in hop nodes in API

tanganellilore committed:
Add peers readonly api and instancelink constraint (#13916)

Add Disconnected link state

introspect_receptor_connections is a periodic
task that examines active receptor connections
and cross-checks it with the InstanceLink info.

Any links that should be active but are not
will be put into a Disconnected state. If
active, it will be in an Established state.

UI - Add hop creation and peers mgmt (#13922)

* add UI for mgmt peers, instance edit and add

* add peer info on detail and bug fix on detail

* remove unused chip and change peer label

* rename lookup, put Instance type disable on edit

---------

Co-authored-by: tanganellilore <lorenzo.tanagnelli@hotmail.it>

fosterseth committed:
Add receptor host identifier to group_vars

Add disconnected link state topology

djyasin committed:
Prevent manual peering of control plane nodes to hop node (#13966)

relrod commented on awx/api/serializers.py:
```diff
@@ -5465,22 +5479,29 @@ def get_percent_capacity_remaining(self, obj):
+        if peers_from_control_nodes and node_type not in (Instance.Types.EXECUTION, Instance.Types.HOP):
+            raise serializers.ValidationError("peers_from_control_nodes can only be enabled for execution or hop nodes.")
+
+        if node_type in (Instance.Types.CONTROL):
```
I would write this as: `if node_type == Instance.Types.CONTROL:`

But not just to be nit-picky.

The line in its current form happens to work, but I don't think it's obvious/intentional how it's working.

This is a peculiarity of Python syntax when defining a tuple with just one element. You need a trailing comma. To someone quickly skimming the code, it _looks_ like you're making a tuple here, but without a trailing comma, you aren't:

```python
>>> a = (Instance.Types.CONTROL)
>>> b = (Instance.Types.CONTROL,)
>>> type(a)
<enum 'Types'>
>>> type(b)
<class 'tuple'>
```

What you're actually checking is if `node_type` is in the tuple that `Instance.Types.CONTROL` is defined as in `models/ha.py` (which gets turned into an enum entry, where the entry itself is a tuple):

```python
    class Types(models.TextChoices):
        CONTROL = 'control', _("Control plane node")
```

So the line happens to work because `"control"` is in that tuple/enum entry. But I don't think that was the intention here when the code was written, I think you wanted to compare against a tuple that you were defining. Which is fine if you are planning to possibly have more entries here in the future (in which case, my request is that you add the comma to the end). But if you're never planning to check against more than one instance type in this conditional, `==` is more clear, and works as expected:

```python
>>> 'control' == Instance.Types.CONTROL
True
>>> 'hop' == Instance.Types.CONTROL
False
```

fosterseth replied:
good catch and great explanation. Will address this

fosterseth commented:
reminder: use string translations on the validationerrors in the serializers

fosterseth committed:
Remove task that enables COPR receptor repo (#14088)

do not pip install receptorctl

thedoubl3j committed:
[hop_node] Validate listener_port is defined for peers (#14056)

add peer listener_port validation and update install bundle if listener_port is defined or not defined.

djyasin committed:
Migration file to set peers_from_control_ nodes to true for existing execution nodes (#14061)

fosterseth committed:
Make peers field optional

djyasin committed:
Marked hop node validation errors for translation (#14116)

thedoubl3j committed:
[hop node] update peer validation logic  (#14132)

fosterseth committed:
Add functional API tests

add tests for calling write_receptor_config

add write_receptor_config test

Do not set default listener_port on control node

djyasin committed:
Hop node AWX Collection Updates (#14153)

Add hop node support to awx collections
- add peers and peers_from_control_nodes fields
- show new node_type "hop"
- add tests for adding hop nodes via collections

Co-authored-by: Seth Foster <fosterseth@users.noreply.github.com>

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -4350,3 +4351,17 @@ def post(self, request, *args, **kwargs):
+    search_fields = ('source', 'target', 'link_state')
+
+
+class PeersDetail(RetrieveAPIView):
```
The plural of "Peers" is kind of weird here - this is showing a single InstanceLink right?

jbradberry replied:
I agree with @AlanCoding (with the additional question in my mind of why they exist)

AlanCoding commented on awx/api/views/root.py:
```diff
@@ -130,6 +130,7 @@ def get(self, request, format=None):
         data['mesh_visualizer'] = reverse('api:mesh_visualizer_view', request=request)
         data['bulk'] = reverse('api:bulk', request=request)
         data['analytics'] = reverse('api:analytics_root_view', request=request)
+        data['peers'] = reverse('api:peers_list', request=request)
```
this is kind of vague to be at /api/v2/, maybe "instance_peers" or "instance_links" would be better

fosterseth replied:
this endpoint isn't used, I think we will drop it altogether

jbradberry replied:
@fosterseth yeah, I think dropping it is the right thing.

fosterseth replied:
removed this endpoint

AlanCoding commented on awx/main/migrations/0186_hop_nodes.py:
```diff
@@ -0,0 +1,53 @@
+
+def set_peers_from_control_nodes_true(apps, schema_editor):
+    Instance = apps.get_model('main', 'Instance')
+    Instance.objects.filter(node_type='execution').update(peers_from_control_nodes=True)
```
should this still be done if it's not a K8S install?

AlanCoding replied:
never mind my comment, that's already being worked on in https://github.com/ansible/awx/pull/14196

djyasin committed:
hop node migration file updates(#14196)

rename migration function set_peers_from_control_nodes_true to automatically_peer_from_control_plane
import settings and only run function if settings.IS_K8S is true
set listener_port for control nodes to None

djyasin committed:
Looking to see if revising the path in the static dir resolves failing ci check.

fosterseth committed:
Add peers_from for reverse peers M2M

use devel receptor-collection

AlanCoding commented:
I think the api-test CI failure could be potentially because this branch needs to be rebased.

AlanCoding commented:
in operator check:

```
Traceback (most recent call last):
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/jobs.py\", line 597, in run
    res = receptor_job.run()
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/receptor.py\", line 318, in run
    res = self._run_internal(receptor_ctl)
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/receptor.py\", line 393, in _run_internal
    res = processor_future.result()
  File \"/usr/lib64/python3.9/concurrent/futures/_base.py\", line 446, in result
    return self.__get_result()
  File \"/usr/lib64/python3.9/concurrent/futures/_base.py\", line 391, in __get_result
    raise self._exception
  File \"/usr/lib64/python3.9/concurrent/futures/thread.py\", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/utils/common.py\", line 1189, in wrapper_cleanup_new_process
    return func(*args, **kwargs)
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/awx/main/tasks/receptor.py\", line 461, in processor
    return ansible_runner.interface.run(
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/ansible_runner/interface.py\", line 210, in run
    r = init_runner(**kwargs)
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/ansible_runner/interface.py\", line 107, in init_runner
    stream_processor = Processor(event_handler=event_callback_handler,
  File \"/var/lib/awx/venv/awx/lib64/python3.9/site-packages/ansible_runner/streaming.py\", line 270, in __init__
    self.artifact_dir = os.path.join(project_artifacts, ident)
  File \"/usr/lib64/python3.9/posixpath.py\", line 90, in join
    genericpath._check_arg_types('join', a, *p)
  File \"/usr/lib64/python3.9/genericpath.py\", line 152, in _check_arg_types
    raise TypeError(f'{funcname}() argument must be str, bytes, or '
TypeError: join() argument must be str, bytes, or os.PathLike object, not 'int'
```

This is fixed with https://github.com/ansible/ansible-runner/pull/1268

fosterseth commented on awx/main/models/ha.py:
```diff
@@ -493,6 +536,13 @@ def on_instance_group_deleted(sender, instance, using, **kwargs):
+
+@receiver(post_save, sender=InstanceLink)
+def on_instancelink_save(sender, instance, using, **kwargs):
+    logger.warning("instancelink saved")
```
remove this on_instancelink_save

djyasin committed:
[hop node] fix failing ci checks on feature_hop-node branch (#14226)

thedoubl3j committed:
[hop node] documentation update in execution_nodes for hop nodes (#14215)

Co-authored-by: Seth Foster <fosterseth@users.noreply.github.com>

fosterseth committed:
receptor python packages

kurokobo committed:
docs: update execution_nodes.md to follow changes for receptor_collection (#14247)

kialam committed:
feature hop node topology updates (#14142)

fosterseth committed:
receptor_python_packages renamed

fosterseth committed:
Allow setting ip_address for execution nodes

fosterseth committed:
optional listener port UI (#14300)

fosterseth commented on awx/main/migrations/0186_hop_nodes.py:
```diff
@@ -0,0 +1,61 @@
+                default=None,
+                help_text='Port that Receptor will listen for incoming connections on.',
+                null=True,
+                validators=[django.core.validators.MinValueValidator(1), django.core.validators.MaxValueValidator(65535)],
```
min value 1024

fosterseth commented on awx/main/tasks/system.py:
```diff
@@ -459,6 +459,30 @@ def execution_node_health_check(node):
+    all_links = InstanceLink.objects.all()
+    active_receptor_conns = mesh_status['KnownConnectionCosts']
+    update_links = []
+    for link in all_links:
```
ignore nodes that are in a deprovisioning state

jbradberry replied:
This all isn't very atomic, leaving the possibility of changes getting clobbered.

fosterseth commented on awx/main/tasks/system.py:
```diff
@@ -459,6 +459,30 @@ def execution_node_health_check(node):
+                link.link_state = InstanceLink.States.ESTABLISHED
+                update_links.append(link)
+        else:
+            if link.link_state is not InstanceLink.States.DISCONNECTED:
```
ignore nodes in adding state

fosterseth commented on awx/main/models/ha.py:
```diff
@@ -464,21 +478,50 @@ def on_instance_group_saved(sender, instance, created=False, raw=False, **kwargs
+        else:
+            control_instances = set(Instance.objects.filter(node_type__in=[Instance.Types.CONTROL, Instance.Types.HYBRID]))
+            if instance.peers_from_control_nodes:
+                if (control_instances & set(instance.peers_from.all())) != set(control_instances):
```
@jbradberry please double check my conditional here

fosterseth commented on awx/main/models/ha.py:
```diff
@@ -464,21 +478,50 @@ def on_instance_group_saved(sender, instance, created=False, raw=False, **kwargs
+                    instance.peers_from.add(*control_instances)
+                    schedule_write_receptor_config()  # keep method separate to make pytest mocking easier
+            else:
+                if set(control_instances) & set(instance.peers_from.all()):
```
remove set

fosterseth committed:
Remove duplicate install bundle on InstanceDetail

fosterseth committed:
Remove extra newlines in install bundle all.yml

fosterseth committed:
Bump migration number 186 to 187

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5368,6 +5374,11 @@ class Meta:
 
+class PeersSerializer(serializers.StringRelatedField):
+    def to_internal_value(self, value):
+        return Instance.objects.get(hostname=value)
```
I'm curious why this was done, instead of making `InstanceSerializer.peers` something like the `source` and `target` fields on `InstanceLinkSerializer`.  It seems to me like doing it this way will result in N+1 query behavior.

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5376,10 +5387,11 @@ class InstanceSerializer(BaseSerializer):
 
     class Meta:
         model = Instance
-        read_only_fields = ('ip_address', 'uuid', 'version')
```
We're allowing `ip_address` to be changed?  Doesn't that have the same problem as changing the `port` value?  A bunch of instances pointing to it will have to have their configs updated.

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5464,37 +5478,62 @@ def get_percent_capacity_remaining(self, obj):
+        if peers_from_control_nodes and node_type not in (Instance.Types.EXECUTION, Instance.Types.HOP):
+            raise serializers.ValidationError(_("peers_from_control_nodes can only be enabled for execution or hop nodes."))
+
+        if node_type == Instance.Types.CONTROL:
```
What about hybrid?  Those are also on the control plane, and presumably would disable this field for the same reason.

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5464,37 +5478,62 @@ def get_percent_capacity_remaining(self, obj):
-            if self.instance.node_type != value:
-                raise serializers.ValidationError("Cannot change node type.")
+        if not self.instance and value not in [Instance.Types.HOP, Instance.Types.EXECUTION]:
+            raise serializers.ValidationError("Can only create execution nodes.")
```
Wording.

fosterseth replied:
now includes hop

fosterseth committed:
Do not change link state if Removing

inspect_established_receptor_connections should
not change link state is current state is Removing.

Other changes:
- rename inspect_execution_nodes to inspect_execution_and_hop_nodes
- Default link state is Adding
- Set min listener_port value to 1024
- inspect_established_receptor_connections now
runs as part of cluster_node_heartbeat task

jbradberry commented on awx/api/templates/instance_install_bundle/requirements.yml:
```diff
@@ -1,4 +1,5 @@
 collections:
-  - name: ansible.receptor
-    version: 1.1.0
+  - name: https://github.com/fosterseth/receptor-collection.git
```
:raised_eyebrow:

jbradberry commented on awx_collection/plugins/modules/instance.py:
```diff
@@ -122,6 +138,12 @@ def main():
+    if peers is not None:
+        new_fields['peers'] = peers
+    if peers is None:
+        peers = ['']
```
Surely the empty list?

fosterseth replied:
fixed

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,331 @@
+        with pytest.raises(IntegrityError):
+            control_instance.peers.add(control_instance)
+
+    @pytest.mark.parametrize('node_type', ['control', 'hop', 'execution'])
```
hybrid

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,331 @@
+            expect=201,
+        )
+
+    @pytest.mark.parametrize('node_type, allowed', [('control', False), ('hop', True), ('execution', True)])
```
hybrid

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,331 @@
+        if peers_from_control_nodes is True, listener_port must an integer
+        Assert that all other combinations are allowed
+        '''
+        Instance.objects.create(hostname='abc', node_type="control")
```
I'd also toss in a (you guessed it) hybrid node, just to make sure that both types do the autolink thing.

fosterseth replied:
the purpose of this test is not autolinking, rather it makes sure listener port is defined if peers_from_control_nodes is enabled. in fact the control node creation is not need so I'm going to remove that line.

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,331 @@
+        for control nodes, peers field should not be
+        modified directly via patch.
+        '''
+        control = Instance.objects.create(hostname='abc', node_type='control')
```
pretty much take as given that it would be good to cover hybrid nodes everywhere else in this module unless it explicitly doesn't make sense (but I can't think of circumstances where that would happen).

fosterseth committed:
Revert "Remove duplicate install bundle on InstanceDetail"

This reverts commit cf5ccf53f4322b49b1009ca13e4f025c30529b30.

jbradberry commented on awx/api/templates/instance_install_bundle/group_vars/all.yml:
```diff
@@ -14,10 +15,22 @@ custom_tls_certfile: receptor/tls/receptor.crt
+{% endfor %}
+{% endif %}
+additional_python_packages:
+  - ansible-runner
```
As I mentioned in the review meeting, this should not be installed on hop nodes.

jbradberry commented on awx/api/urls/urls.py:
```diff
@@ -153,6 +154,7 @@
     re_path(r'^bulk/$', BulkView.as_view(), name='bulk'),
     re_path(r'^bulk/host_create/$', BulkHostCreateView.as_view(), name='bulk_host_create'),
     re_path(r'^bulk/job_launch/$', BulkJobLaunchView.as_view(), name='bulk_job_launch'),
+    re_path(r'^peers/', include(peers_urls)),
```
I don't understand why we are doing top-level 'peers' endpoints.  Peers implies that we are looking in the context of some particular instance, yeah?

Maybe as I get further down I'll see why and this comment will just recommend changing the name to 'links' or something.

fosterseth replied:
dropped this endpoint

fosterseth committed:
Remove duplicate install bundle on InstanceDetail

jbradberry commented on awx/main/access.py:
```diff
@@ -2949,6 +2950,22 @@ def filtered_queryset(self):
         return self.model.objects.filter(workflowjobtemplatenodes__workflow_job_template__in=WorkflowJobTemplate.accessible_pk_qs(self.user, 'read_role'))
 
 
+class PeersAccess(BaseAccess):
```
This should presumably go away along with the peers endpoint

AlanCoding replied:
Right now I'm not seeing any views which are not read-only.

fosterseth replied:
dropping this

jbradberry commented on awx/main/models/ha.py:
```diff
@@ -67,19 +68,28 @@ class InstanceLink(BaseModel):
 
     link_state = models.CharField(
-        choices=States.choices, default=States.ESTABLISHED, max_length=16, help_text=_("Indicates the current life cycle stage of this peer link.")
+        choices=States.choices, default=States.ADDING, max_length=16, help_text=_("Indicates the current life cycle stage of this peer link.")
```
I think this is a mistake, bound up in a misapprehension of what the modeling of the links is there to accomplish, along with several other things I see in this PR.  I'll try to write up my thoughts on this as a top-level comment.

fosterseth replied:
Jeff and I discussed and decided to keep ADDING as the default.

jbradberry commented on awx/main/models/ha.py:
```diff
@@ -67,19 +68,28 @@ class InstanceLink(BaseModel):
     class States(models.TextChoices):
         ADDING = 'adding', _('Adding')
         ESTABLISHED = 'established', _('Established')
+        DISCONNECTED = 'disconnected', _('Disconnected')
```
and this.

jbradberry commented on awx/main/models/ha.py:
```diff
@@ -67,19 +68,28 @@ class InstanceLink(BaseModel):
     class Meta:
         unique_together = ('source', 'target')
+        ordering = ("id",)
+        constraints = [models.CheckConstraint(check=~models.Q(source=models.F('target')), name='source_and_target_can_not_be_equal')]
```
:+1:

jbradberry commented on awx/api/views/instance_install_bundle.py:
```diff
@@ -49,9 +53,9 @@ class InstanceInstallBundle(GenericAPIView):
         instance_obj = self.get_object()
 
-        if instance_obj.node_type not in ('execution',):
+        if instance_obj.node_type not in ('execution', 'hop'):
```
it would be good to make use of the enum objects instead of bare strings, like we do elsewhere

fosterseth replied:
- [ ] use enum where possible

AlanCoding replied:
Yeah, I think this is still an open item but only a minor code style thing.

fosterseth committed:
Remove Disconnected link state

Dynamically flipping from Established
to Disconnected is not the intended
usage of InstanceLink State.

- Link state starts in Adding and becomes
Established once any control node first sees the link
is in the status KnownConnectionCosts

fosterseth committed:
Add listener_port to provision_instance

API changes
- cannot change peers or enable
peers_from_control_nodes on VM deployments
- allow setting ip_address
- use ip_address over hostname in the generated
group_vars/all.yml
- Drop api/v2/peers endpoint

DB changes
- add ip_address unique constraint, but ignore "" entries

Other changes
- provision_instance should take listener_port option

Tests
- test that new controls doesn't disturb other peers
relationships
- test ip_address over hostname

fosterseth committed:
Do not install ansible-runner or podman on hop nodes

fosterseth committed:
Change PeersSerializer to SlugRelatedField

Get rid of PeersSerializer and just use SlugRelatedField,
which should be more a straightforward approach.

Other changes:
- cleanup code related to the already-removed api/v2/peers
endpoint
- add "hybrid" node type into more instance_peers test cases

fosterseth committed:
Fix detecting if peers changed in serializer

Add a check_peers_changed() utility method
to determine if peers in attrs matches
the current instance peers.

Other changes:
- Set ip_address default to "", and do not
allow null.

kurokobo commented:
Created small PR to fix docs for this branch: https://github.com/ansible/awx/pull/14248

fosterseth committed:
Ensure ip_address is empty string

fosterseth committed:
Change username to <username> in inventory

jbradberry commented on awx/main/tasks/receptor.py:
```diff
@@ -731,6 +769,3 @@ def remove_deprovisioned_node(hostname):
     Instance.objects.filter(hostname=hostname).delete()
-
-    # Update the receptor configs for all of the control-plane.
-    write_receptor_config.apply_async(queue='tower_broadcast_all')
```
Why is this being taken out?  Nodes that previously linked to the decommissioned node will need to be updated.

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,368 @@
+
+    @pytest.mark.parametrize('node_type', ['control', 'hybrid'])
+    def test_prevent_peering_to_self(self, node_type):
+        '''
```
this is picky, I know, but we should get away from using triple-single quotes

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,368 @@
+                        user=admin_user,
+                        expect=400 if disallowed else 201,
+                    )
+                    i += 1
```
I would have used `itertools.product` and `enumerate` to avoid manually managing the index number.

fosterseth replied:
good suggestion

jbradberry commented on awx/main/tests/functional/api/test_instance_peers.py:
```diff
@@ -0,0 +1,368 @@
+        control = Instance.objects.create(hostname='abc', node_type=node_type)
+        assert hop in control.peers.all()
+        hop.delete()
+        assert not control.peers.exists()
```
if this test doesn't work, it'll be because `.exists()` uses the queryset results cache populated by the `.all()` above.

fosterseth committed:
Make ip_address read only

Setting a different value for ip_address
and hostname does not work with the current
way we create receptor certs.

fosterseth committed:
Use itertools product instead of nested loop

Make test case cleaner by using itertools product
instead of the triple nested loop

Replace triple single quotes with triple
double quotes

fosterseth committed:
Setup receptor after podman

Might help to install receptor last,
that way when nodes are first connected to the mesh
they already have podman installed and can potentially
run jobs. Otherwise it might be possible for controller
to launch jobs against nodes that aren't fully set up.

fosterseth committed:
No longer assert on receptor_host_identifier

receptor_host_identifier can be left out
of group_vars and will default to the
'ansible_host' variable

fosterseth committed:
Removed unused variable in test_instance_peers

fosterseth committed:
Apply JS formatting from npm prettier

kialam committed:
Add toast and delete modal messaging when removing/adding peers. (#14373)

janorn commented on awx/api/templates/instance_install_bundle/install_receptor.yml:
```diff
@@ -7,14 +7,11 @@
         name: "{{ receptor_user }}"
         shell: /bin/bash
-    - name: Enable Copr repo for Receptor
-      command: dnf copr enable ansible-awx/receptor -y
```
Shouldn't this be removed from previous installed receptors?

fosterseth replied:
sorry, can you clarify the question? The next receptor-collection version will install receptor via these releases by default https://github.com/ansible/receptor/releases

thus the Copr repo is no longer needed going forward

janorn replied:
I'm thinking of upgrades of previous installs.

fosterseth committed:
Use receptor-collection devel

CFSNM commented on awx/api/templates/instance_install_bundle/requirements.yml:
```diff
@@ -1,4 +1,5 @@
-    version: 1.1.0
+  - name: https://github.com/ansible/receptor-collection.git
+    type: git
+    version: main
```
do we need to point this to 2.0.0 version?

AlanCoding commented on tools/docker-compose/bootstrap_development.sh:
```diff
@@ -40,7 +40,7 @@ echo "Admin password: ${DJANGO_SUPERUSER_PASSWORD}"
 awx-manage register_default_execution_environments
 
-awx-manage provision_instance --hostname="$(hostname)" --node_type="$MAIN_NODE_TYPE"
+awx-manage provision_instance --hostname="$(hostname)" --node_type="$MAIN_NODE_TYPE" --listener_port=2222
```
This concerns me a little bit about the expected practice. Does this need to be provided for VM installs in general? I don't get why we didn't need this before, but would now.

AlanCoding replied:
So it seems that this is not required, and as such, it would make the most sense to _not_ set it here, because the VM install scenario that this corresponds to should not ordinarily be setting the listener port.

AlanCoding commented:
Filed a followup issue https://github.com/ansible/awx/issues/14387

fosterseth committed:
Use receptor collection 2.0.0
